File: sdks/ios/Sources/Likerts/ChoiceFeatures.swift
import Foundation

extension Question {
  public var hasOther: Bool { options?.contains { $0.other != nil } == true }

  public func choiceAnswer(selected: [String], otherText: [String: String] = [:]) -> Answer {
    if hasOther {
      let text = Dictionary(uniqueKeysWithValues: (options ?? []).filter {
        $0.other != nil && selected.contains($0.id)
      }.map { ($0.id, otherText[$0.id] ?? "") })
      return .other(selected: selected, otherText: text)
    }
    return type == "single_choice" ? .text(selected.first ?? "") : .choices(selected)
  }

  public func toggledChoices(_ current: [String], option id: String) -> [String] {
    guard let option = options?.first(where: { $0.id == id }) else { return current }
    if type == "single_choice" { return [id] }
    if current.contains(id) { return current.filter { $0 != id } }
    if option.exclusive == true { return [id] }
    return current.filter { id in options?.first(where: { $0.id == id })?.exclusive != true } + [id]
  }

  /// Returns a stable local-feedback code; hosts/renderers supply localized copy.
  public func choiceError(selected: [String], otherText: [String: String] = [:]) -> String? {
    if selected.isEmpty { return required == true ? "required" : nil }
    if Set(selected).count != selected.count || selected.contains(where: { id in
      options?.contains(where: { $0.id == id }) != true
    }) { return "invalid" }
    let exclusive = selected.contains { id in options?.contains { $0.id == id && $0.exclusive == true } == true }
    if exclusive && selected.count != 1 { return "invalid" }
    if type == "single_choice" {
      if selected.count != 1 { return "range" }
    } else if selected.count > (maxSelections ?? options?.count ?? 0)
      || (!exclusive && selected.count < Swift.max(minSelections ?? 0, required == true ? 1 : 0)) {
      return "range"
    }
    let expected = (options ?? []).filter { $0.other != nil && selected.contains($0.id) }
    if otherText.count != expected.count { return "other" }
    for option in expected {
      guard let text = otherText[option.id], !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
        text.unicodeScalars.count <= option.other!.maxLength else { return "other" }
    }
    return nil
  }
}

File: sdks/android/src/main/kotlin/com/likerts/sdk/ChoiceFeatures.kt
package com.likerts.sdk

import kotlinx.serialization.json.*

fun Question.choiceAnswer(selected: List<String>, otherText: Map<String, String> = emptyMap()): JsonElement {
    if (options.any { it.other != null }) return buildJsonObject {
        put("selected", JsonArray(selected.map(::JsonPrimitive)))
        put("otherText", buildJsonObject { options.filter { it.other != null && it.id in selected }.forEach { put(it.id, otherText[it.id] ?: "") } })
    }
    return if (type == "single_choice") JsonPrimitive(selected.firstOrNull() ?: "") else JsonArray(selected.map(::JsonPrimitive))
}

fun Question.toggledChoices(current: List<String>, id: String): List<String> {
    val option = options.find { it.id == id } ?: return current
    if (type == "single_choice") return listOf(id)
    if (id in current) return current - id
    if (option.exclusive == true) return listOf(id)
    return current.filter { selected -> options.find { it.id == selected }?.exclusive != true } + id
}

/** Stable local-feedback code; respondent text belongs to the host's localization. */
fun Question.choiceError(selected: List<String>, otherText: Map<String, String> = emptyMap()): String? {
    if (selected.isEmpty()) return if (required) "required" else null
    if (selected.distinct().size != selected.size || selected.any { id -> options.none { it.id == id } }) return "invalid"
    val exclusive = selected.any { id -> options.any { it.id == id && it.exclusive == true } }
    if (exclusive && selected.size != 1) return "invalid"
    if (type == "single_choice") { if (selected.size != 1) return "range" }
    else if (selected.size > (maxSelections ?: options.size) || (!exclusive && selected.size < maxOf(minSelections ?: 0, if (required) 1 else 0))) return "range"
    val expected = options.filter { it.other != null && it.id in selected }
    if (otherText.size != expected.size) return "other"
    expected.forEach { option -> val text = otherText[option.id] ?: return "other"; if (text.isBlank() || text.codePointCount(0,text.length) > option.other!!.maxLength) return "other" }
    return null
}
